onnxruntime/core/optimizer/embed_layer_norm_fusion.cc

```diff
@@ -512,208 +512,348 @@ Status EmbedLayerNormFusion::ApplyImpl(Graph& graph, bool& modified, int graph_l
-      // Check that input has static shape.
-      auto input_shape = input_ids->Shape();
-      if (input_shape->dim_size() != 2 ||
-          !utils::HasDimValue(input_shape->dim()[0]) ||
-          !utils::HasDimValue(input_shape->dim()[1])) {
-        DEBUG_LOG("Input is expected to have dim value in all dimensions.");
+      // Get input "input_ids" from node.
+      if (!CheckInput(input_ids, logger)) {
+        DEBUG_LOG("Input id is not valid. ");
         continue;
       }
 
-      int64_t batch_size = input_shape->dim()[0].dim_value();
-      int64_t sequence_length = input_shape->dim()[1].dim_value();
-      if (batch_size <= 0 || sequence_length <= 0) {
+      // Get input "mask" from "ReduceSum" node.
+      NodeArg* mask = reduce_sum_node.MutableInputDefs()[0];
+      if (!CheckInput(mask, logger)) {
+        DEBUG_LOG("Mask is not valid. ");
         continue;
       }
 
-      const ONNX_NAMESPACE::TensorProto* position_embed_tensor;
-      if (!graph.GetInitializedTensor(add_input_name, position_embed_tensor)) {
-        DEBUG_LOG("Failed to get initializer tensor.");
+      if (utils::GetTensorShapeFromTensorShapeProto(*(input_ids->Shape())) !=
+          utils::GetTensorShapeFromTensorShapeProto(*(mask->Shape()))) {
+        DEBUG_LOG("Input_ids and mask should have the same shape. ");
         continue;
       }
-      // Tensor shape shall be [batch_size, sequence_length, hidden_size].
-      if (position_embed_tensor->dims_size() != 3 ||
-          position_embed_tensor->dims(0) != batch_size ||
-          position_embed_tensor->dims(1) != sequence_length ||
-          position_embed_tensor->dims(2) != hidden_size) {
-        DEBUG_LOG("Position embedding shape not matched.");
+
+      NodeArg* gamma = layer_norm_node.MutableInputDefs()[1];
+      NodeArg* beta = layer_norm_node.MutableInputDefs()[2];
+      if (gamma->Shape() == nullptr || gamma->Shape()->dim()[0].dim_value() != hidden_size) {
+        DEBUG_LOG("Gamma should be of shape (hidden_size). ");
         continue;
       }
 
-      // Tensor data type should be float or float16.
-      const auto data_type = position_embed_tensor->data_type();
-      if (data_type != ONNX_NAMESPACE::TensorProto_DataType_FLOAT &&
-          data_type != ONNX_NAMESPACE::TensorProto_DataType_FLOAT16) {
-        DEBUG_LOG("Position embedding data type shall be float or float16.");
+      if (beta->Shape() == nullptr || beta->Shape()->dim()[0].dim_value() != hidden_size) {
+        DEBUG_LOG("Beta should be of shape (hidden_size). ");
         continue;
       }
 
-      // The tensor has same data for all batches, and we extract only one batch data as position embedding.
-      position_embedding = ExtractEmbedding(graph, batch_size, sequence_length, hidden_size, position_embed_tensor, modified);
+      // Cast input_ids, segment_ids, and mask to int32 if needed.
+      input_ids = CastToInt32(graph, input_ids, layer_norm_node.GetExecutionProviderType());
```

Comment:
The code of creating EmbedLayerNormation node is quite similar. It is possible to refactor and create a new function to avoid duplicated code.